### PR TITLE
Fragment lighting: implement spot light 

### PR DESCRIPTION
**Pull request by wwylele**
[Here is a testing program](https://gist.github.com/wwylele/3318df2a717d9451722c7a9e5a36b5a8) (adapted from 3ds-examples/fragment-lighting)

Reference: https://www.3dbrew.org/wiki/GPU/Fragment_Lighting (including http://mathb.in/64694)

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/2727)

### CODE REVIEW

File: src/video_core/regs_lighting.h [hunk 1]
```diff
@@ -26,6 +26,14 @@ struct LightingRegs {
         DistanceAttenuation = 16,
     };
 
+    static LightingSampler SpotlightAttenuationSampler(unsigned index) {
+        return (LightingSampler)((unsigned)LightingSampler::SpotlightAttenuation + index);
+    }
+
+    static LightingSampler DistanceAttenuationSampler(unsigned index) {
```
Comment (wwylele) [inline src/video_core/regs_lighting.h]: This is unrelated to spot light implementation. I just want distance attenuation to use the same style. This function will be used for the distance attenuation revise later.

File: src/video_core/regs_lighting.h [hunk 2]
```diff
@@ -26,6 +26,14 @@ struct LightingRegs {
         DistanceAttenuation = 16,
     };
 
+    static LightingSampler SpotlightAttenuationSampler(unsigned index) {
+        return (LightingSampler)((unsigned)LightingSampler::SpotlightAttenuation + index);
```
Comment (wwylele) [inline src/video_core/regs_lighting.h]: I really don't want to use `static_cast` here, which will make this line too long.

Comment (lioncash) [inline src/video_core/regs_lighting.h]: I would still use static cast here. IMO length of a line shouldn't be something which determines what type of cast to use.

**Comment by ruineka** [comment]
I'm not sure if this is in relation to #2723 or not, but seeing Fragment Lighting made me think it might be. I tested the game again on the latest build and it has the same crash. You probably already know this, but I thought I'd say something just to be sure.

**Comment by Subv** [comment]
Do we have a before-after screenshot of any game that uses this?

**Comment by wwylele** [comment]
@ruineka As I already pointed out there, that issue is related to tangent mapping (well, fragment lighting related), but it is unrelated to this PR.

@Subv I am not aware of any game using this (but probably many, for this doesn't produce any log before). I discovered this by comparing registers with 3dbrew documentation.

**Comment by yuriks** [comment]
Reviewed 1 of 5 files at r4, 4 of 4 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/video_core/regs_lighting.h, line 158 at r4](https://reviewable.io:443/reviews/citra-emu/citra/2727#-Kl8SrWuddfVP88Wnwad:-Kl8SrWv_TW9E9Myguud:b-ejww2y) ([raw file](https://github.com/citra-emu/citra/blob/69d9056132ca6d85206ab1185b00aea5d49a5d7c/src/video_core/regs_lighting.h#L158)):*
> ```objc
>         };
> 
>         // Encoded as fixed1.1.11, negated
> ```

"negated" in relation to what? You haven't even describe this as being the spot direction (which I'm assuming it's what it is?) This comment is confusing as is.

---

*[src/video_core/regs_lighting.h, line 241 at r4](https://reviewable.io:443/reviews/citra-emu/citra/2727#-Kl8S9lGPkKeSKyTbmvf:-Kl8S9lGPkKeSKyTbmvg:bg9wwbl) ([raw file](https://github.com/citra-emu/citra/blob/69d9056132ca6d85206ab1185b00aea5d49a5d7c/src/video_core/regs_lighting.h#L241)):*
> ```objc
>                                     config1.disable_spot_light_2, config1.disable_spot_light_3,
>                                     config1.disable_spot_light_4, config1.disable_spot_light_5,
>                                     config1.disable_spot_light_6, config1.disable_spot_light_7};
> ```

These kinds of arrays of bitfields values tend to generate very dreadful code. I think it would be better to add a general mask field covering everything and then do the shift yourself in the functions to check. It's also way less boilerplate.

---

*[src/video_core/renderer_opengl/gl_shader_gen.cpp, line 94 at r5](https://reviewable.io:443/reviews/citra-emu/citra/2727#-Kl8TyAHTKelOMht2j4I:-Kl8TyAHTKelOMht2j4J:bygddbx) ([raw file](https://github.com/citra-emu/citra/blob/09129d8639e47613722283de4052eabca4494175/src/video_core/renderer_opengl/gl_shader_gen.cpp#L94)):*
> ```cpp
>     // this is a dummy field due to lack of the corresponding register
>     state.lighting.lut_sp.enable = true;
> ```

Remove newlines to keep lines grouped

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2727)*

**Comment by wwylele** [comment]
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/video_core/regs_lighting.h, line 158 at r4](https://reviewable.io:443/reviews/citra-emu/citra/2727#-Kl8SrWuddfVP88Wnwad:-Kl9rKMG6k-WNf8N_qDf:b-65e5xj) ([raw file](https://github.com/citra-emu/citra/blob/69d9056132ca6d85206ab1185b00aea5d49a5d7c/src/video_core/regs_lighting.h#L158)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

"negated" in relation to what? You haven't even describe this as being the spot direction (which I'm assuming it's what it is?) This comment is confusing as is.
</blockquote></details>

added more comment

---

*[src/video_core/regs_lighting.h, line 241 at r4](https://reviewable.io:443/reviews/citra-emu/citra/2727#-Kl8S9lGPkKeSKyTbmvf:-Kl9rNMRSH6Bz2DUgI5y:bgvtc19) ([raw file](https://github.com/citra-emu/citra/blob/69d9056132ca6d85206ab1185b00aea5d49a5d7c/src/video_core/regs_lighting.h#L241)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

These kinds of arrays of bitfields values tend to generate very dreadful code. I think it would be better to add a general mask field covering everything and then do the shift yourself in the functions to check. It's also way less boilerplate.
</blockquote></details>

done. was a copy-pasta code from `IsDistAttenDisabled`

---

*[src/video_core/renderer_opengl/gl_shader_gen.cpp, line 94 at r5](https://reviewable.io:443/reviews/citra-emu/citra/2727#-Kl8TyAHTKelOMht2j4I:-Kl9rWXwd2Z87emzjOOz:bb7f2mf) ([raw file](https://github.com/citra-emu/citra/blob/09129d8639e47613722283de4052eabca4494175/src/video_core/renderer_opengl/gl_shader_gen.cpp#L94)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

Remove newlines to keep lines grouped
</blockquote></details>

done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2727)*

**Comment by yuriks** [comment]
:lgtm:

---

Reviewed 5 of 5 files at r7, 4 of 4 files at r8, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/video_core/regs_lighting.h, line 158 at r7](https://reviewable.io:443/reviews/citra-emu/citra/2727#-KlGAft4Pdf-BFyToLn_:-KlGAft4Pdf-BFyToLna:bwcq4mt) ([raw file](https://github.com/citra-emu/citra/blob/f02e86ec9f1946888a9fdcd40b22972af78f836a/src/video_core/regs_lighting.h#L158)):*
> ```objc
>         };
> 
>         // inverse spotlight vector, encoded as fixed1.1.11
> ```

"spotlight direction vector"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2727#-:-KlGBCaDlW2knr3KrHgM:bnfp4nl)*

**Comment by yuriks** [comment]
Reviewed 5 of 5 files at r10, 4 of 4 files at r11, 1 of 1 files at r12.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2727)*

**Comment by Subv** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/video_core/regs_lighting.h, line 134 at r10](https://reviewable.io:443/reviews/citra-emu/citra/2727#-KlLYpR_szoFcrVvWWVs:-KlLYpR_szoFcrVvWWVt:b-7gi0m4) ([raw file](https://github.com/citra-emu/citra/blob/17b4cdcd53cdcd5bab64a335c32965c5ea8aeec3/src/video_core/regs_lighting.h#L134)):*
> ```objc
>                    (config == LightingConfig::Config7);
>         default:
>             UNREACHABLE_MSG("Regs::IsLightingSamplerSupported: Reached "
> ```

Should probably update this log

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2727)*

**Comment by wwylele** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/video_core/regs_lighting.h, line 134 at r10](https://reviewable.io:443/reviews/citra-emu/citra/2727#-KlLYpR_szoFcrVvWWVs:-KlNLU7-gY8m2DXxcYIt:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/17b4cdcd53cdcd5bab64a335c32965c5ea8aeec3/src/video_core/regs_lighting.h#L134)):*
<details><summary><i>Previously, Subv (Sebastian Valle) wrote…</i></summary><blockquote>

Should probably update this log
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2727)*

**Pull request merged**